app/Pasteboard.m

```diff
@@ -0,0 +1,247 @@
+}
+
+static int realloc_to_fit(clip_fd* fd, size_t fit_len) {
+    // (Re)allocate buffer if there's not enough space to fit fit_len
+    if (fit_len <= fd_priv(fd).buffer_cap) {
+        return 0;
+    }
+    if (fit_len > MAXIMAL_BUFFER_CAP) {
+        return 1;
+    }
+
+    size_t size = fd_priv(fd).buffer_cap * 2;
+    if (size == 0) {
+        size = INITIAL_BUFFER_CAP;
+    }
+    while (size < fit_len) size *= 2;
+
+    void *new_buf = realloc(fd_priv(fd).buffer, size);
+    if (new_buf == NULL) {
+        return 1;
+    }
+
+    fd_priv(fd).buffer = new_buf;
+    fd_priv(fd).buffer_cap = size;
+
+    return 0;
+}
+
+// buffer => UIPasteboard
+static int clipboard_wsync(clip_fd *fd) {
+    if (fd_priv(fd).buffer == NULL) {
+        return 0;
+    }
+
+    void *data = fd_priv(fd).buffer;
+    size_t len = fd_priv(fd).buffer_len;
+
+    // FIXME(stek29): This logs "Returning local object of class NSString"
+    // and I have no idea why (or how to fix it)
+    UIPasteboard.generalPasteboard.string = [[NSString alloc]
+                                             initWithBytes:data
+                                             length:len
+                                             encoding:NSUTF8StringEncoding];
+
+    // Reset generation since we've just updated UIPasteboard
+    // note: offset doesn't change
+    fd_priv(fd).generation = 0;
+
+    return 0;
+}
+
+// UIPasteboard => buffer, return len
+static ssize_t clipboard_rsync(clip_fd *fd) {
+    if (fd_priv(fd).buffer != NULL) {
+        free(fd_priv(fd).buffer);
+        fd_priv(fd).buffer = NULL;
+        fd_priv(fd).buffer_cap = 0;
+        fd_priv(fd).buffer_len = 0;
+    }
+
+    size_t len;
+    const void *data = get_data(fd, &len);
+    
+    // Make sure size is still INITIAL_BUFFER_CAP based
+    if (realloc_to_fit(fd, len)) {
+        return _ENOMEM;
+    }
+
+    memcpy(fd_priv(fd).buffer, data, len);
+    fd_priv(fd).buffer_len = len;
+
+    return len;
+}
+
+static int clipboard_poll(clip_fd *fd) {
+    int rv = POLL_WRITE;
+
+    size_t len = 0;
+    if (get_data(fd, &len) != NULL && fd->offset < len) {
+        rv |= POLL_READ;
+    }
+
+    return rv;
+}
+
+static ssize_t clipboard_read(clip_fd *fd, void *buf, size_t bufsize) {
+    size_t length = 0;
+    const char *data = get_data(fd, &length);
+
+    if (data == NULL) {
+        return _EIO;
+    }
+
+    size_t remaining = length - fd->offset;
+    if ((size_t) fd->offset > length)
+        remaining = 0;
+
+    size_t n = bufsize;
+    if (n > remaining)
+        n = remaining;
+
+    if (n != 0) {
+        memcpy(buf, data + fd->offset, n);
+        fd->offset += n;
+    }
+
+    return n;
+}
+
+static ssize_t clipboard_write(clip_fd *fd, const void *buf, size_t bufsize) {
+    size_t new_len = fd->offset + bufsize;
+    size_t old_len = fd_priv(fd).buffer_len;
+
+    if (realloc_to_fit(fd, new_len)) {
+        return _ENOMEM;
+    }
+
+    // fill the hole between new offset and old len
```

Comment:
Should fill in the hole between old_len and new_len, not old_len and fd->offset

Reply:
nope, it should feel between old_len and fd->offset since between offset and new_len is where buf gets copied to -- there's no point in memset'ing and then copying buffer over it.

Reply:
Ah, gotcha.